lib/streamlit/delta_generator_singletons.py

```diff
@@ -0,0 +1,131 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from contextvars import ContextVar, Token
+from typing import TYPE_CHECKING, Callable, Generic, TypeVar
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+    from streamlit.elements.lib.dialog import Dialog
+    from streamlit.elements.lib.mutable_status_container import StatusContainer
+
+"""
+The main purpose of this module (right now at least) is to avoid a dependency
+cycle between streamlit.delta_generator and some elements.
+"""
+
+
+main_dg: DeltaGenerator | None = None
+sidebar_dg: DeltaGenerator | None = None
+event_dg: DeltaGenerator | None = None
+bottom_dg: DeltaGenerator | None = None
+
+
+def get_main_dg() -> DeltaGenerator:
+    if main_dg is None:
+        raise RuntimeError("main_dg is not initialized")
+    return main_dg
+
+
+def get_sidebar_dg() -> DeltaGenerator:
+    if sidebar_dg is None:
+        raise RuntimeError("sidebar_dg is not initialized")
+    return sidebar_dg
+
+
+def get_event_dg() -> DeltaGenerator:
+    if event_dg is None:
+        raise RuntimeError("event_dg is not initialized")
+    return event_dg
+
+
+def get_bottom_dg() -> DeltaGenerator:
+    if bottom_dg is None:
+        raise RuntimeError("bottom_dg is not initialized")
+    return bottom_dg
+
+
+# The dg_stack tracks the currently active DeltaGenerator, and is pushed to when
+# a DeltaGenerator is entered via a `with` block. This is implemented as a ContextVar
+# so that different threads or async tasks can have their own stacks.
```

Comment:
nit: maybe put this as proper function docstring to get_default_dg_stack

Reply:
good point, I have moved it to be proper docstrings